File: solutions/_1856_maximum_subarray_min_product/solution.py
from typing import List


class Solution:
    def maxSumMinProduct(self, nums: List[int]) -> int:
        res = 0
        stack = []
        prefix = [0]

        for num in nums:
            prefix.append(prefix[-1] + num)

        for i, num in enumerate(nums):
            newStart = i
            while stack and stack[-1][1] > num:
                start, val = stack.pop()
                total = prefix[i] - prefix[start]
                res = max(res, val * total)
                newStart = start
            stack.append((newStart, num))

        for start, val in stack:
            total = prefix[len(nums)] - prefix[start]
            res = max(res, val * total)

        return res % (10**9 + 7)


# https://leetcode.com/problems/maximum-subarray-min-product/

File: solutions/_1856_maximum_subarray_min_product/solution.cpp
#include <vector>
using namespace std;

class Solution
{
public:
    int maxSumMinProduct(vector<int> &nums)
    {
        int n = nums.size();
        long long res = 0;
        vector<long long> prefix(n + 1);
        for (int i = 0; i < n; i++)
        {
            prefix[i + 1] = prefix[i] + nums[i];
        }
        vector<int> stack;

        for (int i = 0; i <= n; i++)
        {
            while (!stack.empty() && (i == n || nums[stack.back()] > nums[i]))
            {
                int j = stack.back();
                stack.pop_back();
                int k = stack.empty() ? -1 : stack.back();
                res = max(res, nums[j] * (prefix[i] - prefix[k + 1]));
            }
            stack.push_back(i);
        }

        return res % 1000000007;
    }
};

/*
https://leetcode.com/problems/maximum-subarray-min-product/
*/